warehouse/internal/repo/staging.go

```diff
@@ -0,0 +1,288 @@
+package repo
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/utils/timeutil"
+	"github.com/rudderlabs/rudder-server/warehouse/internal/model"
+	warehouseutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+)
+
+const tableName = warehouseutils.WarehouseStagingFilesTable
+
+const tableColumns = `
+	id,
+    location,
+    source_id,
+    destination_id,
+    schema,
+    error,
+    status,
+    first_event_at,
+    last_event_at,
+    total_events,
+    created_at,
+    updated_at,
+    metadata,
+    workspace_id 
+`
+
+// StagingFiles is a repository for inserting and querying staging files.
+type StagingFiles struct {
+	DB  *sql.DB
+	Now func() time.Time
+
+	once sync.Once
+}
+
+type metadataSchema struct {
+	UseRudderStorage      bool   `json:"use_rudder_storage"`
+	SourceBatchID         string `json:"source_batch_id"`
+	SourceTaskID          string `json:"source_task_id"`
+	SourceTaskRunID       string `json:"source_task_run_id"`
+	SourceJobID           string `json:"source_job_id"`
+	SourceJobRunID        string `json:"source_job_run_id"`
+	TimeWindowYear        int    `json:"time_window_year"`
+	TimeWindowMonth       int    `json:"time_window_month"`
+	TimeWindowDay         int    `json:"time_window_day"`
+	TimeWindowHour        int    `json:"time_window_hour"`
+	DestinationRevisionID string `json:"destination_revision_id"`
+}
+
+func metadataFromStagingFile(stagingFile *model.StagingFile) metadataSchema {
+	return metadataSchema{
+		UseRudderStorage:      stagingFile.UseRudderStorage,
+		SourceBatchID:         stagingFile.SourceBatchID,
+		SourceTaskID:          stagingFile.SourceTaskID,
+		SourceTaskRunID:       stagingFile.SourceTaskRunID,
+		SourceJobID:           stagingFile.SourceJobID,
+		SourceJobRunID:        stagingFile.SourceJobRunID,
+		TimeWindowYear:        stagingFile.TimeWindow.Year(),
+		TimeWindowMonth:       int(stagingFile.TimeWindow.Month()),
+		TimeWindowDay:         stagingFile.TimeWindow.Day(),
+		TimeWindowHour:        stagingFile.TimeWindow.Hour(),
+		DestinationRevisionID: stagingFile.DestinationRevisionID,
+	}
+}
+
+func (m *metadataSchema) SetStagingFile(stagingFile *model.StagingFile) {
+	stagingFile.UseRudderStorage = m.UseRudderStorage
+	stagingFile.SourceBatchID = m.SourceBatchID
+	stagingFile.SourceTaskID = m.SourceTaskID
+	stagingFile.SourceTaskRunID = m.SourceTaskRunID
+	stagingFile.SourceJobID = m.SourceJobID
+	stagingFile.SourceJobRunID = m.SourceJobRunID
+	stagingFile.TimeWindow = time.Date(m.TimeWindowYear, time.Month(m.TimeWindowMonth), m.TimeWindowDay, m.TimeWindowHour, 0, 0, 0, time.UTC)
+	stagingFile.DestinationRevisionID = m.DestinationRevisionID
+}
+
+func (repo *StagingFiles) init() {
+	repo.once.Do(func() {
+		if repo.Now == nil {
+			repo.Now = timeutil.Now
+		}
+	})
+}
```

Comment:
can you explain the rational behind this special function ?
Why not assign the function to `repo.Now` at time of its creation ?

Reply:
The idea behind this is that we can mock time for testing purposes.

We can set `Now` during tests to return a static/predictable time. If Now is not set, it defaults to `timeutil.Now`.

The overall idea behind `init` is lazy initialisation. It is a helpful pattern used from many components in the standard library. We are using the repo struct itself for configuration. Construction implicitly happens when the first method is called.